src/rnnlm/sample_a_word.cc

```diff
@@ -0,0 +1,432 @@
+          if (it != vocab_.end()) {
+            history.push_back(it->second);
+          } else {
+            std::cout << "OOV found in history: " << tokens[i] << std::endl;
+          }
+        }
+        assert (history.size() == order_current - 1);
+        log_prob = std::stof(tokens[0]);
+        backoff_weight = std::stof(tokens.back());
+        probs_pair = std::make_pair(log_prob, backoff_weight);
+        probs_[order_current - 1][history].insert({word, probs_pair});
+        continue;
+      }
+      // case2: no backoff_weights
+      if (tokens.size() == order_current + 1 && (tokens.back() == "</s>") && tokens[0] != "ngram") {
+        unordered_map<std::string, int32>::iterator it = vocab_.find(tokens.back());
+        if (it != vocab_.end()) {
+          word = it->second;
+        } 
+        int32 len_hist = tokens.size() - 2; // exclude the word and log-prob
+        HistType history;
+        assert (len_hist > 0);
+        for (int32 i = 1; i < len_hist + 1; i++) {
+          unordered_map<std::string, int32>::iterator it = vocab_.find(tokens[i]);
+          if (it != vocab_.end()) {
+            history.push_back(it->second);
+          } else {
+            std::cout << "OOV found in history: " << tokens[i] << std::endl;
+          }
+        }
+        assert (history.size() == order_current - 1);
+        log_prob = std::stof(tokens[0]);
+        backoff_weight = 0; // backoff_weight in log space should be 1 (no backoff)
+        probs_pair = std::make_pair(log_prob, backoff_weight);
+        probs_[order_current - 1][history].insert({word, probs_pair});
+        continue;
+      }
+    } else if (order_current == ngram_order_) { // read probs of order N
+      if (tokens.size() > 2) {
+        std::string word_s = tokens.back();
+        unordered_map<std::string, int32>::iterator it = vocab_.find(word_s);
+        if (it != vocab_.end()) {
+          word = it->second;
+        }
+        int32 len_hist = tokens.size() - 2; // exclude the word and log-prob
+        HistType history;
+        assert (len_hist > 0);
+        for (int32 i = 1; i < len_hist + 1; i++) {
+          unordered_map<std::string, int32>::iterator it = vocab_.find(tokens[i]);
+          if (it != vocab_.end()) {
+            history.push_back(it->second);
+          } else {
+            std::cout << "OOV found in history: " << tokens[i] << std::endl;
+          }
+        }
+        log_prob = std::stof(tokens[0]);
+        backoff_weight = 0; // backoff_weight in log space should be 1 (no backoff)
+        probs_pair = std::make_pair(log_prob, backoff_weight);
+        probs_[order_current - 1][history].insert({word, probs_pair});
+        continue;
+      }
+    }
+  }
+  std::cout << "Finish reading ARPA-format file." << std::endl;
+}
+
+float NgramModel::GetProb(int32 order, const int32 word, const HistType& history) {
+  float prob = 0.0;
+  auto it = probs_[order - 1].find(history);
```

Comment:
no auto.

OK, please check ALL my comments on your previous pull requests. You're making the same mistakes again.